google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc_test.cc

```diff
@@ -0,0 +1,702 @@
+bool operator==(FakeRequest const& lhs, FakeRequest const& rhs) {
+  return lhs.key == rhs.key;
+}
+
+bool operator==(FakeResponse const& lhs, FakeResponse const& rhs) {
+  return lhs.key == rhs.key && lhs.value == rhs.value;
+}
+
+using AsyncReaderWriter = ::google::cloud::pubsublite_internal::MockAsyncReaderWriter<
+    FakeRequest,
+    FakeResponse>;
+
+using AsyncReadWriteStreamReturnType =
+std::unique_ptr<AsyncStreamingReadWriteRpc<FakeRequest, FakeResponse>>;
+
+using ResumableAsyncReadWriteStream = std::unique_ptr<
+    ResumableAsyncStreamingReadWriteRpc<FakeRequest, FakeResponse>>;
+
+const Status kFailStatus = Status(StatusCode::kUnavailable, "Unavailable");
+const FakeResponse kBasicResponse = FakeResponse{"key0", "value0_0"};
+const FakeResponse kBasicResponse1 = FakeResponse{"key0", "value0_1"};
+const FakeRequest kBasicRequest = FakeRequest{"key0"};
+
+class ResumableAsyncReadWriteStreamingRpcTest : public ::testing::Test {
+ protected:
+  ResumableAsyncReadWriteStreamingRpcTest() : backoff_policy_{
+      std::make_shared<StrictMock<MockBackoffPolicy>>()}, stream_{
+      MakeResumableAsyncStreamingReadWriteRpcImpl<FakeRequest, FakeResponse>(
+          [this]() { return retry_policy_factory_.Call(); },
+          backoff_policy_,
+          [this](std::chrono::duration<double> t) { return sleeper_.Call(t); },
+          [this]() { return stream_factory_.Call(); },
+          [this](AsyncReadWriteStreamReturnType stream) {
+            return initializer_.Call(std::move(stream));
+          })} {
+  }
+
+  StrictMock<MockFunction<future<void>(std::chrono::duration<double>)>>
+      sleeper_;
+  StrictMock<MockFunction<AsyncReadWriteStreamReturnType()>> stream_factory_;
+  StrictMock<MockFunction<std::unique_ptr<MockRetryPolicy>()>>
+      retry_policy_factory_;
+  StrictMock<MockFunction<future<StatusOr<AsyncReadWriteStreamReturnType>>(
+      AsyncReadWriteStreamReturnType)>>
+      initializer_;
+  std::shared_ptr<StrictMock<MockBackoffPolicy>> backoff_policy_;
+  ResumableAsyncReadWriteStream stream_;
+
+  void InitializerBehavior(unsigned int times) {
+    EXPECT_CALL(initializer_, Call).Times(times).WillRepeatedly([](
+        AsyncReadWriteStreamReturnType stream) {
+      return make_ready_future(
+          StatusOr<AsyncReadWriteStreamReturnType>(std::move(stream)));
+    });
+  }
+
+  void NoRetryInvocation(unsigned int times) {
+    // Return(ByMove(...)) results in runtime error on times > 1 ?
+    EXPECT_CALL(retry_policy_factory_,
+                Call).Times(times).WillRepeatedly([]() {
+      return absl::make_unique<StrictMock<
+          MockRetryPolicy>>();
+    });
+  }
+
+  void StandardSingleRetryPolicy() {
+    auto mock_retry_policy =
+        absl::make_unique<StrictMock<MockRetryPolicy>>();
+    EXPECT_CALL(*mock_retry_policy, IsExhausted).WillOnce(Return(false));
+    EXPECT_CALL(*mock_retry_policy, OnFailure(_)).WillOnce(Return(true));
+    EXPECT_CALL(retry_policy_factory_, Call)
```

Comment:
These calls should be split based on where they are called in the code.

If you want to stall an action (such as the stream initialization) so you can split up these actions, you can do (for that example):

promise<StatusOr<AsyncReadWriteStreamReturnType>> initializer_promise;
promise<AsyncReadWriteStreamReturnType> leaked_stream;

EXPECT_CALL(initializer_, Call).WillOnce([&](AsyncReadWriteStreamReturnType stream){
  leaked_stream.set_value(std::move(stream));
});

// Set other expectations that happen before initialization

auto start = stream->Start();

initializer_promise.set_value(leaked_stream.get());